Use `testcloud` domain API v2

frantisekz opened the pull request:
Refactors domain configuration interface between tmt and testcloud, to allow greater flexibility and superior api stability when making changes and additions in the future.

Apart from that, doesn't expose support for (will come as a separate PR):
- tpm emulation
- uefi
- cpu core count setting (more advanced topologies to follow https://tmt.readthedocs.io/en/stable/spec/hardware.html#cpu TBD)

testcloud counterparts: 
- https://pagure.io/testcloud/c/6f0942cb3ba31ff0927ee8b151218e8dc8b34d7d?branch=master
- https://pagure.io/testcloud/pull-request/155
- https://pagure.io/testcloud/c/7ee6dc839a46c72db6d23bcd2e057c2de6fe5ea8?branch=master
- ...

(compatible testcloud built at): https://copr.fedorainfracloud.org/coprs/frantisekz/testcloud-wip/

kkaarreell commented:
Hi @frantisekz , I would like to test this code but may I ask you to rebase first? For proper testing using my test I need the multi-host related code so it would help me a lot if this would be up to date. Thank you.

frantisekz commented:
> Hi @frantisekz , I would like to test this code but may I ask you to rebase first? For proper testing using my test I need the multi-host related code so it would help me a lot if this would be up to date. Thank you.

Sure, done!

frantisekz commented:
@thrix both addressed, thanks!

happz commented:
@frantisekz any chance you might be willing to delay your PR to 1.27 to be merged together with https://github.com/teemtee/tmt/pull/1316?

My main concern is the new options added by your patch, `--cpus`, `--tpm`, and `--uefi`. If we do indeed get rid of the EPEL8 limitation in 1.27, the HW requirements PR could get merged and I would invest my time into helping transition testcloud plugin to use the `--hardware` key/option & follow the existing specification. Obviously, these new features are something you wish to see supported by the plugin, same here, yet I'm worried about the number of options duplicating what's already on the verge of being ready. EPEL8 was a long-time blocker, but since things started to move very recently, we might actually get rid of the limitation in 1.27, unblocking the consolidation.

I already updated the Beaker plugin in https://github.com/teemtee/tmt/pull/1316/, to "understand" common HW requirements as specified in `hardware` key. There is a CLI option, `--hardware`, mirroring the `hardware` key, allowing the same degree of CLI extensibility as `--disk`, `--memory` or `--cpus`, but the same language accepted by the `hardware` key: `--hardware "cpu.processors>=4"`, `--hardware "disk[0].size >= 100 GiB"`, `--hardware "tpm.version == 2"`.

I understand it would be yet another delay, but it would make the Universe a much better and tidier place :) Switching the plugin's implementation to one used by the `beaker` and `artemis` plugins would be a huge boost to tmt's power of running a plan on top of many different infrastructures.

lukaszachy commented:
I'm sorry this didn't get into 1.26, see above for reasons.

frantisekz commented:
> @frantisekz any chance you might be willing to delay your PR to 1.27 to be merged together with #1316?
> 

Sure sure, no problem.

> My main concern is the new options added by your patch, `--cpus`, `--tpm`, and `--uefi`. If we do indeed get rid of the EPEL8 limitation in 1.27, the HW requirements PR could get merged and I would invest my time into helping transition testcloud plugin to use the `--hardware` key/option & follow the existing specification. 
>
> Obviously, these new features are something you wish to see supported by the plugin, same here, yet I'm worried about the number of options duplicating what's already on the verge of being ready. EPEL8 was a long-time blocker, but since things started to move very recently, we might actually get rid of the limitation in 1.27, unblocking the consolidation.
> 
> I already updated the Beaker plugin in #1316, to "understand" common HW requirements as specified in `hardware` key. There is a CLI option, `--hardware`, mirroring the `hardware` key, allowing the same degree of CLI extensibility as `--disk`, `--memory` or `--cpus`, but the same language accepted by the `hardware` key: `--hardware "cpu.processors>=4"`, `--hardware "disk[0].size >= 100 GiB"`, `--hardware "tpm.version == 2"`.

I'll remove exposure of `--cpus`, `--tpm`, and `--uefi` from this PR and make a separate one for it with the migration to a more standardized approach. From the testcloud development perspective, the far more important is the migration to the new api, the exposure of other params is a nice "side effect" (I do realize this is more important to tmt consumers). IMO, it'd also make more sense from commit/PR perspective to have these two split.

wdyt @happz ?

happz commented:
> > My main concern is the new options added by your patch, `--cpus`, `--tpm`, and `--uefi`. If we do indeed get rid of the EPEL8 limitation in 1.27, the HW requirements PR could get merged and I would invest my time into helping transition testcloud plugin to use the `--hardware` key/option & follow the existing specification.
> > Obviously, these new features are something you wish to see supported by the plugin, same here, yet I'm worried about the number of options duplicating what's already on the verge of being ready. EPEL8 was a long-time blocker, but since things started to move very recently, we might actually get rid of the limitation in 1.27, unblocking the consolidation.
> > I already updated the Beaker plugin in #1316, to "understand" common HW requirements as specified in `hardware` key. There is a CLI option, `--hardware`, mirroring the `hardware` key, allowing the same degree of CLI extensibility as `--disk`, `--memory` or `--cpus`, but the same language accepted by the `hardware` key: `--hardware "cpu.processors>=4"`, `--hardware "disk[0].size >= 100 GiB"`, `--hardware "tpm.version == 2"`.
> 
> I'll remove exposure of `--cpus`, `--tpm`, and `--uefi` from this PR and make a separate one for it with the migration to a more standardized approach.
>
> From the testcloud development perspective, the far more important is the migration to the new api, the exposure of other params is a nice "side effect" (I do realize this is more important to tmt consumers). IMO, it'd also make more sense from commit/PR perspective to have these two split.
> 
> wdyt @happz ?


TBH, that would be the best. I'm not opposed to having these options per se, but I'd like to see them:

1. beeing the same across plugins that support `hardware`, i.e. `-h virtual --memory 8Gib` vs `-h beaker --ram 8GiB` would be a defeat, and
2. following the specification, i.e. `--cpu-processors` (or `--cpu.processors`?) rather than `--cpus` as `cpus` is not defined by https://tmt.readthedocs.io/en/stable/spec/hardware.html#cpu, `--memory` instead of `--ram`, `--tpm-version=2.0", and so on.

I'm pretty sure we can amend explicit "shortcuts" like `--memory` with whatever is in a plan's `hardware` field. My current idea would be to use `and` to merge these together

```yaml
provision:
  hardware:
    cpu:
      processors: >= 6
    boot:
      method: uefi
```

```console
$ tmt run -a provision --memory "> 8GiB" --disk "120 GiB" ...
```

could result in the following set of constraints:

```yaml
and:
  - cpu:
      processors: >= 6
    boot:
      method: uefi
  - memory: > 8GiB
  - disk:
      size: 120 GiB
```

I bet there can be other ways, but in general, no problem with having these options. Adding them to `artemis`, `beaker`, and `virtual` plugins in a standalone set of patches would make sense to me. And don't worry, I'm perfectly fine spending my time on this, after all, I'm making things more complicated now, it's absolutely fitting for me to get rid of all the wrinkles in all those plugins later :)

psss commented:
> From the testcloud development perspective, the far more important is the migration to the new api...

Makes a complete sense to me as well to cover api changes first in a separate PR.

happz commented:
@frantisekz I noticed you dropped the extra features, is your PR stable now, can I try to refactor arch/memory/disk to use the new HW requirements code on top of this patch? I'd start with these existing ones, then we can add tpm, CPU core, and the rest.

happz commented on tmt/steps/provision/testcloud.py:
```diff
@@ -1,4 +1,5 @@
 # coding: utf-8
+# mypy: disable-error-code="name-defined"
```
I'm not sure we should take this road, disabling a rule like this globally can easily prevent mypy from spotting actual errors. Is this about the `global *` objects imported from `testcloud` package, the amount of possible `type: ignore[...]` needed to silence them all?

frantisekz replied:
Resolved, it certainly doesn't look nice though...

happz replied:
I suppose these stem out of not being declared on the module level, unlike `libvirt` and `testcloud`. mypy is unable to realize our code would call `import_testcloud()` that would set these "global" names. Would you mind declaring them on the module level, I suppose `DomainConfiguration: Any`, without any value, should introduce the name well enough for mypy to not require `name-defined` anymore?

frantisekz replied:
Yep, that did the trick! Thanks!

(Took me a while before I'd realized I need to import the Any type... :) )

happz commented on tmt/steps/provision/testcloud.py:
```diff
@@ -468,44 +422,84 @@ def start(self) -> None:
         self.instance_name = self._tmt_name()
+
+        # Prepare DomainConfiguration object before Instance object
+        self._domain = DomainConfiguration(self.instance_name)
```
Please, add a declaration similar to `_image` and `_instance`, it's easier to keep track of the attributes of the plugin instance.

happz commented on tmt/steps/provision/testcloud.py:
```diff
@@ -468,44 +422,84 @@ def start(self) -> None:
+        self.legacy_os = testcloud.util.needs_legacy_net(self._image.name)
+
+        # Is this a CoreOS?
+        self.coreos = bool(re.search('coreos|rhcos', self.image.lower()))
```
Do these three names have to be attributes of `self`? They do not seem to be used by any other method, a local scope should be sufficient then.

frantisekz commented:
> @frantisekz I noticed you dropped the extra features, is your PR stable now, can I try to refactor arch/memory/disk to use the new HW requirements code on top of this patch? I'd start with these existing ones, then we can add tpm, CPU core, and the rest.

Yep, it's stable now. I had to look into CoreOS regression which is now resolved (I'll submit a newer testcloud build into the WIP copr).

> 
> Edit: after reading the patch, I'm really looking forward to this! Code's now much cleaner and easier to navigate for me, and I can clearly see where testcloud and HW requirements would interact. This would be the first time the HW requirements code ported from Artemis would meet a plugin actually creating a guest, so far we were matching requirements with pre-existing instance types, and while in theory everything is clear, who knows how it all turns out :) But I'm confident we're going to make it work :)

Thanks! There's still a long road ahead, but the baseline was reaaaally low :) . There is also bunch of stuff pending cleanup once I can remove back/forward compatibility from testcloud.

Thanks for the review items, I believe I'd resolved everything but the mypy which is being worked on right now.

frantisekz commented:
Okay, all should be addressed and a new testcloud (0.9.2-40) has been built in the COPR.

frantisekz commented on tmt/steps/provision/testcloud.py:
```diff
@@ -352,13 +317,6 @@ def _guess_image_url(self, name: str) -> str:
-        with open(DOMAIN_TEMPLATE_FILE, 'w') as template:
-            template.write(DOMAIN_TEMPLATE)
-
     def wake(self) -> None:
```
FYI, this method isn't ported yet. I'd probably do that in a following PR as it'd require splitting start to be able to execute code responsible for constructing DomainConfiguration to be able to call it there too.

If you prefer otherwise, I can handle it as a part of this PR. Also, I hadn't figured out how to actually trigger this codepath, am I blind or is it just a placeholder fn?

psss replied:
The `wake()` method is used to connect to a running instance when using an existing tmt run, for example like this:

```
tmt run provision -h virtual
tmt run --last login
tmt run --last finish
```

It seems that everything needed to connect to the instance is working nicely so I see no blocker here.

frantisekz replied:
Thanks for the way to call the code, will address in a future PR targeting 1.28.

psss replied:
Great, thanks!

happz commented on tmt/steps/provision/testcloud.py:
```diff
@@ -468,44 +423,90 @@ def start(self) -> None:
+
+        # Prepare DomainConfiguration object before Instance object
+        self._domain = DomainConfiguration(self.instance_name)  # type: ignore[name-defined]
+        self._domain.memory_size = self.memory * 1024
```
How is `self.disk` used? I see `self._instance.disk_size = self.disk` being removed, just like `self._instance.ram = self.memory` is, but while the latter is replaced by the line above, the former does not seem to have a new form in the patch.

thrix replied:
@frantisekz can you check?

psss reviewed (changes requested):
Looks good and works nicely! Thanks! Added just one minor suggestion. Plus, I guess, we should also bump the `testcloud` require to make sure the new api is installed on the user system, right?

frantisekz commented on tmt/steps/provision/testcloud.py:
```diff
@@ -42,11 +54,33 @@ def import_testcloud() -> None:
+            )
     except ImportError:
         raise ProvisionError(
             "Install 'testcloud' to provision using this method.")
```
Yep, good point, updated!

frantisekz commented:
> Looks good and works nicely! Thanks! Added just one minor suggestion. Plus, I guess, we should also bump the `testcloud` require to make sure the new api is installed on the user system, right?

Updated, feedback addressed, thanks!

psss commented:
> Updated, feedback addressed, thanks!

Nice, thanks!

Hint: During the pull request review it's better to [add new commits rather then amending existing ones](https://tmt.readthedocs.io/en/latest/contribute.html#pull-requests). It makes it easier to see what actually changed, especially when the force-push is combined with a rebase.

frantisekz commented:
> Hint: During the pull request review it's better to [add new commits rather then amending existing ones](https://tmt.readthedocs.io/en/latest/contribute.html#pull-requests). It makes it easier to see what actually changed, especially when the force-push is combined with a rebase.

Yeah, sorry for that, I probably took it as too trivial no do it the proper way around, will pay attention to it the next time.

psss commented:
> Yeah, sorry for that, I probably took it as too trivial no do it the proper way around, will pay attention to it the next time.

Sure, no problemo.

psss commented:
/packit test

psss commented:
/packit build
